cpp/src/request_mem.cpp

```diff
@@ -0,0 +1,196 @@
+}
+
+void RequestMem::memGetCallback(void* request, ucs_status_t status, void* arg)
+{
+  Request* req = reinterpret_cast<Request*>(arg);
+  ucxx_trace_req_f(req->getOwnerString().c_str(), nullptr, request, "memGet", "memGetCallback");
+  return req->callback(request, status);
+}
+
+void RequestMem::request()
+{
+  ucp_request_param_t param = {.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+                                               UCP_OP_ATTR_FIELD_FLAGS |
+                                               UCP_OP_ATTR_FIELD_USER_DATA,
+                               .flags     = UCP_AM_SEND_FLAG_REPLY,
+                               .datatype  = ucp_dt_make_contig(1),
+                               .user_data = this};
+
+  void* request = nullptr;
+
+  std::visit(data::dispatch{
+               [this, &request, &param](data::MemPut memPut) {
+                 param.cb.send = memPutCallback;
+                 request       = ucp_put_nbx(_endpoint->getHandle(),
+                                       memPut._buffer,
+                                       memPut._length,
+                                       memPut._remoteAddr,
+                                       memPut._rkey,
+                                       &param);
+               },
+               [this, &request, &param](data::MemGet memGet) {
+                 param.cb.send = memGetCallback;
+                 request       = ucp_get_nbx(_endpoint->getHandle(),
+                                       memGet._buffer,
+                                       memGet._length,
+                                       memGet._remoteAddr,
+                                       memGet._rkey,
+                                       &param);
+               },
+               [](auto) { throw std::runtime_error("Unreachable"); },
+             },
+             _requestData);
+
+  std::lock_guard<std::recursive_mutex> lock(_mutex);
+  _request = request;
+}
+
+static void logPopulateDelayedSubmission() {}
+
+void RequestMem::populateDelayedSubmission()
+{
+  bool terminate =
+    std::visit(data::dispatch{
+                 [this](data::MemPut memPut) {
+                   if (_endpoint->getHandle() == nullptr) {
+                     ucxx_warn("Endpoint was closed before message could be sent");
+                     Request::callback(this, UCS_ERR_CANCELED);
+                     return true;
+                   }
+                   return false;
+                 },
+                 [this](data::MemGet memGet) {
+                   if (_worker->getHandle() == nullptr) {
+                     ucxx_warn("Endpoint was closed before message could be received");
+                     Request::callback(this, UCS_ERR_CANCELED);
+                     return true;
+                   }
+                   return false;
+                 },
+                 [](auto) -> decltype(terminate) { throw std::runtime_error("Unreachable"); },
+               },
```

Comment:
Both of these branches are the same

Reply:
They're not, the warning message are different. The general purpose is anyway to ensure we treat all supported types with their own branches, and `throw` on unsupported types, thus providing strong typing.